trieste/utils/pareto.py

```diff
@@ -0,0 +1,309 @@
+                self.bounds.append(atleast2d(idx_lb), atleast2d(idx_ub))
+            # Reject test:
+            elif self._is_test_required((lb + jitter) < self.front):
+                # Cell can not be discarded: calculate the size of the cell
+                dc_dist = cell[1] - cell[0]
+                hc = BoundedVolumes(lb, ub)
+
+                # Only divide when it is not an unit cell
+                # and the volume is above the approx. threshold
+                if tf.reduce_any(dc_dist > 1) and tf.reduce_all(
+                    (hc.size()[0] / total_size) > self.threshold
+                ):
+                    # Divide the test cell over its largest dimension
+                    edge_size, idx = tf.reduce_max(dc_dist), tf.argmax(dc_dist)
+                    edge_size1 = int(tf.round(tf.cast(edge_size, dtype=tf.float32) / 2.0))
+                    edge_size2 = edge_size - edge_size1
+
+                    # Store divided cells
+                    ub = tf.identity(cell[1])
+
+                    ub = tf.unstack(ub)
+                    ub[idx] = ub[idx] - edge_size1
+                    ub = tf.stack(ub)
+                    dc.append((tf.identity(cell[0]), ub))
+
+                    lb = tf.identity(cell[0])
+                    lb = tf.unstack(lb)
+                    lb[idx] = lb[idx] + edge_size2
+                    lb = tf.stack(lb)
+                    dc.append((lb, tf.identity(cell[1])))
+            # else: cell can be discarded
+
+    def bounds_2d(self):
+        """
+        Computes the cells covering the non-dominated region
+        for the specific case of only two objectives.
+        Assumes the Pareto set has been sorted in ascending order on the first objective.
+        This implies the second objective is sorted in descending order.
+        """
+        outdim = self.data.observations.shape[1]
+        assert outdim == 2
+
+        pf_idx = tf.argsort(self.front, axis=0)
+        pf_ext_idx = tf.concat(
+            (
+                atleast2d(tf.zeros(outdim, dtype=tf.int32)),
+                atleast2d(pf_idx + 1),
+                atleast2d(tf.ones(outdim, dtype=tf.int32) * self.front.shape[0] + 1),
+            ),
+            axis=0,
+        )
+
+        for i in range(pf_ext_idx[-1, 0]):
+            self.bounds.append(
+                atleast2d(tf.constant([i, 0], dtype=tf.int32)),
+                atleast2d(tf.constant([i + 1, pf_ext_idx[-i - 1, 1].numpy()], dtype=tf.int32)),
+            )
+
+    def hypervolume(self, reference: tf.Tensor) -> tf.Tensor:
+        """
+        Autoflow method to calculate the hypervolume indicator
+        The hypervolume indicator is the volume of the dominated region.
+        :param reference: reference point to use
+            Should be equal or bigger than the anti-ideal point of the Pareto set
+            For comparing results across runs the same reference point must be used
+        :return: hypervolume indicator (the higher the better)
```

Comment:
```suggestion
        Autoflow method to calculate the hypervolume indicator
        The hypervolume indicator is the volume of the dominated region.

        :param reference: reference point to use
            Should be equal or bigger than the anti-ideal point of the Pareto set
            For comparing results across runs the same reference point must be used
        :return: hypervolume indicator (the higher the better)
```
paragraph needed for docs to render correctly

Reply:
applies to all docstrings